src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs

```diff
@@ -0,0 +1,536 @@
+            {
+                this.logger.LogTrace("(-)[INVALID_NETWORK]");
+                throw new ConsensusException("INVALID_NETWORK");
+            }
+
+            this.SetPeerTip(LocalPeerId, consensusTip.HashBlock);
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// A new list of headers are presented from a given peer, the headers will try to be connected to the chain.
+        /// Headers that are interesting (i.e may extend our consensus tip) will marked to download there full blocks.
+        /// </summary>
+        /// <remarks>
+        /// The headers are assumed to be consecutive in order. 
+        /// </remarks>
+        /// <param name="networkPeerId">The network id that is presenting this headers.</param>
+        /// <param name="headers">The list of headers to connect to the chain tree.</param>
+        /// <returns>Indicators of what blocks need to be downloaded.</returns>
+        public ConnectedHeaders ConnectNewHeaders(int networkPeerId, List<BlockHeader> headers)
+        {
+            Guard.NotNull(headers, nameof(headers));
+            this.logger.LogTrace("({0}:{1},{2}.{3}:{4})", nameof(networkPeerId), networkPeerId, nameof(headers), nameof(headers.Count), headers.Count);
+
+            if (!this.chainedHeadersByHash.ContainsKey(headers[0].HashPrevBlock))
+            {
+                this.logger.LogTrace("(-)[HEADER_COULD_NOT_CONNECT]");
+                throw new ConnectHeaderException();
+            }
+
+            List<ChainedHeader> newChainedHeaders = this.CreateNewHeaders(headers);
+
+            this.SetPeerTip(networkPeerId, headers.Last().GetHash());
+
+            if (newChainedHeaders == null)
+            {
+                this.logger.LogTrace("(-)[NO_NEW_HEADER]");
+                return new ConnectedHeaders();
+            }
+
+            ChainedHeader earliestNewHeader = newChainedHeaders.First();
+            ChainedHeader latestNewHeader = newChainedHeaders.Last();
+
+            ConnectedHeaders connectedHeaders = null;
+
+            bool isAssumedValidEnabled = this.consensusSettings.BlockAssumedValid != null;
+            bool isWithinCheckpoints = this.consensusSettings.UseCheckpoints && (earliestNewHeader.Height <= this.checkpoints.GetLastCheckpointHeight()); ;
+
+            if (isWithinCheckpoints || isAssumedValidEnabled)
+            {
+                ChainedHeader currentChainedHeader = latestNewHeader;
+
+                while (currentChainedHeader != earliestNewHeader)
+                {
+                    if (currentChainedHeader.HashBlock == this.consensusSettings.BlockAssumedValid)
+                    {
+                        this.logger.LogDebug("Chained header '{0}' represents an assumed valid block.", currentChainedHeader);
+
+                        connectedHeaders = this.HandleAssumedValidHeader(currentChainedHeader, latestNewHeader);
+                        break;
```

Comment:
So what would happen if the break was not here? We would continue processing and if there was no checkpoint before the assume valid block we were just finished the loop as if the break was there and we would exit the method online 173.

However, if there was checkpoint, then we would call handle checkpoint header and it would give us new result, but that is probably fine. Because either it was the last checkpoint is, in which case we would request to download all headers we have which is fine regardless of whether there is any assume valid header; or it would be not the last checkpoint, in which case we would request a download everything up until the checkpoint, but not something after that before assume valid block. That would be a mistake, but not a significant one, because we would download that once we reached the final checkpoint anyway.

So it seems that removing the break is the right thing to do. Can someone verify?

Reply:
Okay there is a minor problem with line 253 where we actually mark the chain headers as required for downloading. Therefore we should consider the return value of this method when we are calling handling of checkpoint below, if the break is removed. Download to in the result should be the higher block header from both calls.

@noescape00 we should also create a test for this very specific situation, because it is hard to get right.

Reply:
If we remove the break, which I think we should, we need to put a big comment here explaining the situation.

Reply:
Although it seems that is very rare a case, we don't actually control assume valid setting from commandline and the user may provide such a hash that leads to the situation and so I think we should handle it correctly.

Reply:
So the solution seems to be to pass information about whether we are before last checkpoint to the handle assume valid header method.